server/src/main/java/org/opensearch/index/seqno/LocalCheckpointTracker.java

```diff
@@ -134,11 +142,28 @@ public synchronized void markSeqNoAsPersisted(final long seqNo) {
         markSeqNo(seqNo, persistedCheckpoint, persistedSeqNo);
     }
 
+    /**
+     * Updates the processed sequence checkpoint to the given value.
+     *
+     * This method is only used for segment replication since indexing doesn't
+     * take place on the replica allowing us to avoid the check that all sequence numbers
+     * are consecutively processed.
+     *
+     * @param seqNo the sequence number to mark as processed
+     */
+    public synchronized void fastForwardProcessedSeqNo(final long seqNo) {
+        advanceMaxSeqNo(seqNo);
+        if (shouldUpdateSeqNo(seqNo, processedCheckpoint, persistedCheckpoint) == false) {
```

Comment:
It's a little weird to mix a synchronized method with AtomicLongs, but I think the way you'd write this to be safe is:

```
final long currentProcessedCheckpoint = processedCheckpoint.get();
if (shouldUpdateSeqNo(seqNo, currentProcessedCheckpoint, persistedCheckpoint) == false) {
    return;
}
processedCheckpoint.compareAndSet(currentProcessedCheckpoint, seqNo);
```

Maybe the `synchronized` prevents it in practice (which raises the question of why an AtomicLong is being used...), but in theory the value of processedCheckpoint could change in between the `shouldUpdateSeqNo` check and the `compareAndSet`, which would defeat the purpose of the compare-and-set check.

Reply:
You mean because the method is synchronized, it's redundant to use AtomicLong since it's already thread safe? I was trying to keep to the format markSeqNo followed (similarly called within a synchronized method with atomic parameters). 

I'll make the change as mentioned - thanks for the suggestion - it makes sense esp in context of compareAndSet

Reply:
> You mean because the method is synchronized, it's redundant to use AtomicLong since it's already thread safe?

Yeah, they are both mechanisms for handling concurrency, so it is a bit weird to mix them. You generally use the Atomic* variants when you need limited functionality (like compare-and-set or increment-and-get) because it can give you much better performance versus the more course-grained locks of synchronizing the method.

A cursory look suggests to me that all mutations of `processedCheckpoint` happen in synchronized methods so it could likely be replaced with a volatile `long`. I'm definitely not suggesting you refactor the existing code though.